Add A/B to RevisionDiff revision descriptions #7626

gerhardol opened the pull request:
Discussion in #4154
Replacing #7626, already reviewed and approved, just changing A->a, B->b
This gives yet another chance to suggest changes

## Proposed changes
Add _(a)_ and _(b)_ to some revision descriptions in RevisionDiff

## Screenshots

### Before

![image](https://user-images.githubusercontent.com/6248932/72210720-1c37b700-34c0-11ea-84ce-c95b1db0ad25.png)

![image](https://user-images.githubusercontent.com/6248932/72210725-31144a80-34c0-11ea-8e55-67047c91fe42.png)

### After

![image](https://user-images.githubusercontent.com/6248932/73126311-4d7cb080-3fb1-11ea-80ef-662813b6a75b.png)

![image](https://user-images.githubusercontent.com/6248932/73126318-608f8080-3fb1-11ea-9c2a-2058f36ecc0a.png)

## Test methodology <!-- How did you ensure quality? -->
manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -210,18 +210,14 @@ private string DescribeRevision([CanBeNull] ObjectId objectId, int maxLength = 0
         /// </summary>
         /// <returns>A description of the selected parent</returns>
-        private string DescribeSelectedParentRevision()
+        private string DescribeRevision(IEnumerable<GitRevision> parents)
```
```suggestion
        private string DescribeRevision(List<GitRevision> parents)
```

and replace `parents.Count()` with `parents.Count`

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -749,18 +745,19 @@ private void resetFileToToolStripMenuItem_DropDownOpening(object sender, EventAr
 
-            var parents = DiffFiles.SelectedItemParents.Distinct();
+            var parents = DiffFiles.SelectedItemParents.ToList();
             if (parents.Count() != 1 || !CanResetToRevision(parents.FirstOrDefault()))
```
```suggestion
            if (parents.Count != 1 || !CanResetToRevision(parents.FirstOrDefault()))
```

gerhardol commented:
Applied RussKies suggestions
I am not happy at all about lower case (a)...

![image](https://user-images.githubusercontent.com/6248932/73020424-e1217600-3e25-11ea-802d-34f2d885c204.png)

RussKie commented:
> I am not happy at all about lower case (a)...
> 
> ![image](https://user-images.githubusercontent.com/6248932/73020424-e1217600-3e25-11ea-802d-34f2d885c204.png)

Hmm that doesn't look right to me either.
I suppose I wasn't clear in communicating my thoughts. I was thinking of something like this:

```
-- Diff with a/6a34728274: feature/bla
-- Diff with Commit Index
-- Diff with Working directory
```

mstv commented:
I'd like to put everything in (although it might be not completely visible):
```
-- Diff a/6a347282: feature/bla with b/12345678: master commit
-- Diff a/Commit Index with b/abcdef01: bla
-- Diff a/Working directory with b/87654321: blub
```

![grafik](https://user-images.githubusercontent.com/36601201/73121768-f27d9600-3f7d-11ea-92f6-7b5567a04f52.png)

I'd omit the parens in the context menu items:
`First: b/59d118a9: WaitForPendingOperations in FormCommitTes...`

In addition, the shortened commit hash could be added to the DiffView.
before:
![grafik](https://user-images.githubusercontent.com/36601201/73121832-c0b8ff00-3f7e-11ea-9144-a6fff28284c2.png)

after:
```
diff --git a/UnitTests/CommonTestUtils/AsyncTestHelper.cs b/UnitTests/CommonTestUtils/AsyncTestHelper.cs
index 328374f10..278d18c10 100644
--- 851b482c: a/UnitTests/CommonTestUtils/AsyncTestHelper.cs
+++ 59d118a9: b/UnitTests/CommonTestUtils/AsyncTestHelper.cs
@@ -2,6 +2,7 @@
 using System.Diagnostics;
```

gerhardol commented:
Updated to what I hoped is the final version...

I did not get the admittance I requested here:
https://github.com/gitextensions/gitextensions/pull/7626#issuecomment-574353852
You were saved by my merge mistake!

> I suppose I wasn't clear in communicating my thoughts. 

https://github.com/gitextensions/gitextensions/pull/7626#issuecomment-576122671
Changed to @mstv latest suggestion above, no refactoring neither too

> I'd like to put everything in (although it might be not completely visible):

Some thoughts for later, this is still just tuning current presentation.

> -- Diff a/6a347282: feature/bla with b/12345678: master commit

For now, selected/b revision is fixed and not included. (except for maybe Combined diff)

If #4154 I prefer to have something that makes it easier to distinguish the odd commit with a separate selected/b.

```Diff b/6a347282: feature/bla from a/12345678: master commit```

> In addition, the shortened commit hash could be added to the DiffView

Later if changed, a/b and revisions are explained in the header already.

```diff --git a/UnitTests/CommonTestUtils/AsyncTestHelper.cs b/UnitTests/CommonTestUtils/AsyncTestHelper.cs
index 328374f10..278d18c10 100644
```
Color code the header?

gerhardol commented:
@msftbot merge in 1 day

mstv commented:
> index 328374f10..278d18c10 100644

This information is something git-internal -- I'd say _private_. It looks like commit hashs but that's not the case.
So we should rather remove it.

pmiossec commented:
It should be the blob hash. We could verify it easily doing `git cat-file -p`.

But I didn't know we could remove it in the diff.
we really could?

vbjay commented:
Yes. Its object hashes.   I think it's --no-index

On Sat, Jan 25, 2020, 3:56 PM Philippe Miossec <notifications@github.com>
wrote:

> It should be the blob hash. We could verify it easily doing git cat-file
> -p.
>
> But I didn't know we could remove it in the diff.
> we really could?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7666?email_source=notifications&email_token=AA3WDML4Z5GLWXOSU5AWZH3Q7SROVA5CNFSM4KJZI4I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ5FI2Y#issuecomment-578442347>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMJBHUBOBHDL4KI5E5LQ7SROVANCNFSM4KJZI4IQ>
> .
>

vbjay commented:
Nope. I was wrong. --no-index does something else.

On Sat, Jan 25, 2020, 3:59 PM Jay Asbury <vbjay.net@gmail.com> wrote:

> Yes. Its object hashes.   I think it's --no-index
>
> On Sat, Jan 25, 2020, 3:56 PM Philippe Miossec <notifications@github.com>
> wrote:
>
>> It should be the blob hash. We could verify it easily doing git cat-file
>> -p.
>>
>> But I didn't know we could remove it in the diff.
>> we really could?
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gitextensions/gitextensions/pull/7666?email_source=notifications&email_token=AA3WDML4Z5GLWXOSU5AWZH3Q7SROVA5CNFSM4KJZI4I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ5FI2Y#issuecomment-578442347>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AA3WDMJBHUBOBHDL4KI5E5LQ7SROVANCNFSM4KJZI4IQ>
>> .
>>
>

gerhardol commented:
> So we should rather remove it.

> But I didn't know we could remove it in the diff.

This discussion is OT for merging now, still relevant here

We create the patch, parsing Git diff output, sure we could remove or reformat the output 

Proposal

```--- a/851b482c: UnitTests/CommonTestUtils/AsyncTestHelper.cs
+++ b/59d118a9: UnitTests/CommonTestUtils/AsyncTestHelper.cs```
